Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -1342,12 +1340,12 @@ private void updateInformationFromServer(DocumentEvent e) {
         final int eventOffset = e.getOffset();
         final int deltaLength = e.getNewLength() - e.getOldLength();
         for (PluginNavigationRegion region : regions) {
-          if (region.offset <= eventOffset && eventOffset <= region.offset + region.length) {
-            region.length += deltaLength;
-          }
-          else if (region.offset >= eventOffset) {
+          if (eventOffset <= region.offset) {
             region.offset += deltaLength;
           }
+          else if (region.offset < eventOffset && eventOffset <= region.offset + region.length) {
+            region.length += deltaLength;
```

Comment:
1. I agree, typing in a region most probably changes its resolution, unless it's typing a letter and backspacing it 100 ms later. So, yes, any intersecting regions should be deleted.

2. Highlighting is visible, and it would be less ideal to just remove highlighting regions. For example when user changes the name of a variable in its declaration, we have a special highlighting region, and it should be nice to keep this highlighting, not just turn it into black-on-white, because any other variable name will have the same highlighting.